### ISSUE TITLE
When adding a new data item: Uncaught TypeError: Cannot read property 'packageModel' of undefined

### ISSUE THREAD

**Issue by amoeba** [NCEAS/metacatui#66]
I'm getting this in my console: `DataItemView.js?v=2.0.0:216 Uncaught TypeError: Cannot read property 'packageModel' of undefined`

Steps to reproduce:

1. Point your local MetacatUI at `dev.nceas` and edit urn:uuid:1d36d337-795a-4782-bca0-74ca6f6b194b
2. Once the page loads, click "Add files" and add a file
3. Observe the error message

The error is popping up in `DataItemView.getParentDataPackage`, specifically:

```{js}
                    
                    // A nested package    
                    } else {
                        return MetacatUI.rootDataPackage.where({id: parentResourceMapId})[0];
                        
                    }
```

I don't yet understand this part of the codebase so I'm sharing this here in case others might understand what the fix is.

**Comment by csjx** [comment]
Alright, so I thought I had tracked this down, but only really fooled myself into thinking so. The multiple triple statement entries look to be fixed, however, we're still getting multiple aggregations in the resource map, and I've narrowed this down to a few calls to `IndexedFormula.statementsMatching()` in `DataPackage.serialize()`where they don't return the correct number of statements based on visual inspection of the RDF.  Using a fresh RDF document and loading it with `rdf.parse()`, this call works as advertised.  However, when we make changes to the subjects and/or objects in the triple statements, we get trouble.  So, I'm wondering if, while we are making the changes to the `NamedNode` attributes, they aren't being reflected in the underlying graph, and so the searches are failing.  So my next step is to look at that angle to see what we need to do to properly modify the graph.  This is ridiculously tedious trying to keep track of all of the UUIDs inside the RDF. But I'm whining. 😬

**Comment by csjx** [comment]
Just to update where I'm at on this issue, I'm trying to understand why we have differing results from `IndexedFormula.statementsMatching()` when searching by `subject, predicate` vs `subject, predicate, object`, as in:
```
> var st = aggregationObjStatements[0] // statements with 59cf#aggregation as the object
undefined
> st.subject.value
"https://cn.dataone.org/cn/v2/resolve/urn%3Auuid%3A5e77a2e7-e928-4907-9151-e54ea0f16c87"
> st.predicate.value
"http://www.openarchives.org/ore/terms/describes"
> st.object.value
"https://cn.dataone.org/cn/v2/resolve/urn%3Auuid%3A88140820-a5c2-43d2-8bfb-336ed55e59cf#aggregation"
> this.dataPackageGraph.statementsMatching(st.subject)
[Statement, Statement, Statement, Statement, Statement]
> this.dataPackageGraph.statementsMatching(st.subject, st.predicate)
[Statement]
> this.dataPackageGraph.statementsMatching(st.subject, st.predicate, st.object)
[] // WTF
```
given these existing triples:

Subject | Predicate | Object
------- | --------- | ------
0 | http://xmlns.com/foaf/0.1/name | Christopher Jones
0 | http://www.w3.org/1999/02/22-rdf-syntax-ns#type | http://purl.org/dc/terms/Agent
https://cn.dataone.org/cn/v2/resolve/urn%3Auuid%3A5e77a2e7-e928-4907-9151-e54ea0f16c87 | http://purl.org/dc/elements/1.1/creator | 0
https://cn.dataone.org/cn/v2/resolve/urn%3Auuid%3A5e77a2e7-e928-4907-9151-e54ea0f16c87 | http://purl.org/dc/terms/modified | 2017-04-05T21:37:42.210Z
https://cn.dataone.org/cn/v2/resolve/urn%3Auuid%3A5e77a2e7-e928-4907-9151-e54ea0f16c87 | http://www.w3.org/1999/02/22-rdf-syntax-ns#type | http://www.openarchives.org/ore/terms/ResourceMap
https://cn.dataone.org/cn/v2/resolve/urn%3Auuid%3A5e77a2e7-e928-4907-9151-e54ea0f16c87 | http://www.openarchives.org/ore/terms/describes | https://cn.dataone.org/cn/v2/resolve/urn%3Auuid%3A88140820-a5c2-43d2-8bfb-336ed55e59cf#aggregation
https://cn.dataone.org/cn/v2/resolve/urn%3Auuid%3A5e77a2e7-e928-4907-9151-e54ea0f16c87 | http://purl.org/dc/terms/identifier | urn:uuid:5e77a2e7-e928-4907-9151-e54ea0f16c87
https://cn.dataone.org/cn/v2/resolve/urn%3Auuid%3A88140820-a5c2-43d2-8bfb-336ed55e59cf#aggregation | http://www.openarchives.org/ore/terms/isDescribedBy | https://cn.dataone.org/cn/v2/resolve/urn%3Auuid%3A5e77a2e7-e928-4907-9151-e54ea0f16c87

The call to `this.statementsMatching(subject, predicate, object)` is failing inside `IndexedFormula.remove()`, and so we aren't able to update triple statements where the old aggregation is the object.  Still working on it.

**Comment by csjx** [comment]
I think I've finally tracked down the last of our resource map woes when it comes to updating packages multiple times.  The main issue had to do with `subject` or `object` `Node`s inside of `Statement`s in the `IndexedFormula` index being update by reference, whereas our calls were assuming we were updating by value.  So, when iterating through statements to `IndexedFormula.remove()`, the internal call to `statementsMatching()` would fail and the statement would no longer be removed.  We then add the statment back in with the new pid for the package, and the repeat statements would start accumulating in the graph.  I dealt with this by cloning the statement nodes so we weren't removing by reference, but rather by value.  Hopefully this will stick. Closing until further notice :)
